streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl;
+
+public final class ProcessorContextUtils {
```

Comment:
Thanks!

Reply:
I'm trying to phrase this is a less discouraging way, since I agree this is a nice & clever fix: 

is this meant to be the actual, final solution to this problem, or just a temporary workaround to unblock us without the need for a KIP? This just "happens" to work out nicely because we know the mocks are actually returning a StreamsMetricsImpl as well. What happens if we need to add/access more complicated processor context functionality in the inner state stores? I say "inner" because the caching layers for example also perform a cast to InternalProcessorContext.

I completely agree with proceeding with this, so don't take that question as a challenge to this PR. Just wondering where this leaves us going forward. Should we accept (and therefore enforce) that state stores can't have caching enabled in unit tests?

Reply:
Uh, no, that was an oversight :) I might as well fix it in this PR, so I'll not merge just now. Thanks for pointing it out.

I think where it leaves us, in a nutshell, is that state stores should never cast the context, not as long as their `init()` method is a public interface. This class is a way to collect all the casting-related sins and put them all in one place where we can keep an eye on them.

Casting doesn't always indicate a design failure, but in this case, it 100% does. I'd hope that everything in this class gets designed away and we can delete the class. I'd flip the table and do it right now, but I don't want to block the unit-testability of the PAPI behind a bunch of KIPs. I'll make tickets to fix the stuff that needs to be fixed.

In contrast, note that all the Processor implementations that power our DSL are very much _not_ for public use, so for them, casting to InternalProcessorContext is fair game. Although, I _have_ found and/or fixed a fair number of bugs following from casting in the processors as well. So, although, it's not a contract violation, it still may not be a good idea.

Reply:
Ok, after looking into the internals, I can see that caching is like logging, it only makes sense in the context of being in a "real" KafkaStreams environment, so we don't need to support it for unit tests. I've added more explicit exceptions when this condition is violated, as well as more documentation to the demo tests.

Reply:
I totally agree. I missed to add that I really like this as a temporary fix. IMO, the root cause is that we pass in internal information in a public API which we should not do at all. I have already an idea how to fix it. Let's discuss that somewhere else.